### PR TITLE
Feature/scmi 62625/validation code molecule

### PR DESCRIPTION
**Pull request by Alverd04**
## MOLECULE/VALIDATIONCODE
**TASK**: [SCMI-62625](https://jira.scmspain.com/browse/SCMI-62625) 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Only
- [x] Refactor

### Description, Motivation and Context
Validation Code es the molecule for the atom Pin Input. This molecule adds more functionality to the pinInput and provides the user more tools to work with.

### Screenshots - Animations
Desktop
image
Mobile
image

### CODE REVIEW

File: components/atom/pinInput/test/index.test.js [hunk 1]
```diff
@@ -560,7 +560,8 @@ describe('AtomPinInput', () => {
       } = store
 
       // Then
-      expect(Object.getOwnPropertyNames(others).length).to.equal(0)
+      console.log(Object.getOwnPropertyNames(others).length)
```
Comment (andresin87) [inline components/atom/pinInput/test/index.test.js]: 👀

File: components/atom/pinInput/test/index.test.js [hunk 2]
```diff
@@ -593,7 +594,8 @@ describe('AtomPinInput', () => {
         elements,
         ...others
       } = store
-      expect(Object.getOwnPropertyNames(others).length).to.equal(0)
+
+      expect(Object.getOwnPropertyNames(others).length).to.equal(1)
```
Comment (andresin87) [inline components/atom/pinInput/test/index.test.js]: extract ValueType and let the expect to 0

File: components/atom/pinInput/test/index.test.js [hunk 3]
```diff
@@ -618,7 +620,7 @@ describe('AtomPinInput', () => {
         elements,
         ...others
       } = store
-      expect(Object.getOwnPropertyNames(others).length).to.equal(0)
+      expect(Object.getOwnPropertyNames(others).length).to.equal(1)
```
Comment (andresin87) [inline components/atom/pinInput/test/index.test.js]: same

File: components/molecule/validationCode/test/index.test.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    // Then
+    const input = getByDisplayValue(props.value)
+    expect(input.value).to.exist
+
+    // And
+    // When
+    const deleteButton = getByText(props.deleteButtonTextLabel)
+    userEvents.click(deleteButton)
+    // Then
+    sinon.assert.called(spy)
```
Comment (andresin87) [inline components/molecule/validationCode/test/index.test.js:98]: 👏

Comment (Alverd04) [reply components/molecule/validationCode/test/index.test.js:98]: 🥰🥰🥰🥰

File: components/atom/pinInput/src/config.js [hunk 5]
```diff
@@ -27,3 +27,22 @@ export const valueChecker = ({length = 1, mask}) => (value = '') => {
...
+  } else if (
+    typeof defaultValue === 'object' &&
+    defaultValue instanceof Array
+  ) {
+    valueType = 'array'
+  } else {
+    valueType = 'string'
+  }
+  return valueType
+}
```
Comment (nistadev) [inline components/atom/pinInput/src/config.js]: I didn't try it, but I think this approach should work and is a lot simpler:
```suggestion
export const getValueType = ({ value, defaultValue }) => {
  const val = value || defaultValue

  let valueType = typeof val

  if (typeof val === "object" && val instanceof Array) {
    valueType = 'array'
  }

  return valueType
}
```

File: components/molecule/validationCode/src/validationCode.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+} from './config'
+import useControlledState from '@s-ui/react-hooks/lib/useControlledState'
+import {forwardRef, useMemo} from 'react'
+
+const normalizeValue = value => {
+  if (value === undefined) {
+    return value
+  }
+  const result = `${typeof value === 'string' ? value.split('') : value}`
+  return result
```
Comment (nistadev) [inline components/molecule/validationCode/src/validationCode.js]:
```suggestion
  return `${typeof value === 'string' ? value.split('') : value}`
```

File: components/atom/pinInput/src/config.js (hunk 5 above)
Comment (Alverd04) [reply components/atom/pinInput/src/config.js]: works!

File: components/molecule/validationCode/src/validationCode.js (hunk 6 above)
Comment (Alverd04) [reply components/molecule/validationCode/src/validationCode.js]: applied!

File: components/atom/pinInput/src/PinInput.js [hunk 7]
```diff
@@ -42,10 +42,10 @@ const PinInput = forwardRef(
...
 
     useUpdateEffect(() => {
-      dispatch(
-        pinInputActions.setValue({innerValue: value ? value.split('') : []})
-      )
-    }, [value, dispatch])
+      const innerValue =
+        typeof value === 'string' ? value.split('') : value || []
+      dispatch(pinInputActions.setValue({innerValue}))
+    }, [`${value}`, dispatch])
```
[the remark this answers is not in the document]
Comment (andresin87) [reply components/atom/pinInput/src/PinInput.js:48]: value can be array.

When it happens, the its instance is new each time:
```js
[1,2,3] !== [1,2,3]
```

so you want to compare when each value of the array changes...
```js
`${[1,2,3]}` === `${[1,2,3]}`
```
...to do not fire setValues each render

**Review by nistadev — APPROVED** [review]

File: components/atom/pinInput/test/index.test.js [hunk 8]
```diff
@@ -581,18 +581,20 @@ describe('AtomPinInput', () => {
 
       // When
       const hook = setupReducerEnvironment(args)
-
       // Then
       const [store] = hook.result.current
+      console.log(store)
```
Comment (oriolpuig) [inline components/atom/pinInput/test/index.test.js]: Ok, this is a test but, we need this console.log()?

**Review by oriolpuig — APPROVED** [review]

File: components/atom/pinInput/test/index.test.js (hunk 8 above)
Comment (andresin87) [reply components/atom/pinInput/test/index.test.js:586]: XD, removed in: https://github.com/SUI-Components/sui-components/pull/1665

**Pull request merged**